### PR TITLE
WebForms integration

### PR DESCRIPTION
**Pull request by boydc7**
Instrumentation for WebForms integration.  Implemented as @lucaspimentel originally suggested via an IHttpModule that can be used via a simple web.config entry like so:

```
<system.webServer>
   <modules>
      <add name="AspNetWebFormsHttpModule" type="Datadog.Trace.ClrProfiler.Integrations.AspNetWebFormsHttpModule" />
...
```

Which allows deployment without the native profiler (i.e. no IL manipulation, install, etc.).  Auto instrumentation basically injects that IHttpModule into an HttpApplication pipeline at runtime on Init of the application.

@lucaspimentel biggest up-front question/thing to review that I'd wonder if is correct is the Span decorations (i.e. setting the ResourceName, Host, etc.) to the correct values.  I wasn't able to find a way that any RouteData would ever be sent from a WebForms site, but if we should implement it nonetheless (making certain assumptions about what the things are/would be), or if you know of how to get RouteData sent from a forms site...I'm all ears.  Should be an easy hook-up if so.

This has a very soft dependency on the demo repo branch of the same name ([here](https://github.com/DataDog/dd-trace-demo/tree/chad/webforms)), where the .Managed.sln solution links to the webforms demo project there - no reason to hold up either branch for either in my opinion, as everything works fine without it being in sync.

There is basically a copy of the demo webforms proj in the Samples folder here used currently for the integration test off webforms, ideally we'd work towards leveraging demo projects, data, samples, etc. for our integration tests (which would provide much more value for both us and customers IMHO), but we don't have to do that now naturally.

### CODE REVIEW

File: Datadog.Trace.Managed.sln.DotSettings [hunk 1]
```diff
@@ -22,15 +25,20 @@
...
+	<s:Boolean x:Key="/Default/CodeStyle/CodeFormatting/CSharpFormat/USE_INDENT_FROM_VS/@EntryValue">False</s:Boolean>
+	<s:Boolean x:Key="/Default/CodeStyle/CodeFormatting/CSharpFormat/WRAP_AFTER_DECLARATION_LPAR/@EntryValue">True</s:Boolean>
 	<s:Boolean x:Key="/Default/CodeStyle/CodeFormatting/CSharpFormat/WRAP_AFTER_INVOCATION_LPAR/@EntryValue">True</s:Boolean>
 	<s:String x:Key="/Default/CodeStyle/CodeFormatting/CSharpFormat/WRAP_ARGUMENTS_STYLE/@EntryValue">CHOP_IF_LONG</s:String>
 	<s:String x:Key="/Default/CodeStyle/CodeFormatting/CSharpFormat/WRAP_CHAINED_METHOD_CALLS/@EntryValue">CHOP_IF_LONG</s:String>
 	<s:Boolean x:Key="/Default/CodeStyle/CodeFormatting/CSharpFormat/WRAP_LINES/@EntryValue">False</s:Boolean>
+	<s:String x:Key="/Default/CodeStyle/CodeFormatting/CSharpFormat/WRAP_PARAMETERS_STYLE/@EntryValue">CHOP_IF_LONG</s:String>
 	<s:Boolean x:Key="/Default/CodeStyle/CppUseAuto/AutoWasChosen/@EntryValue">False</s:Boolean>
+	<s:Boolean x:Key="/Default/CodeStyle/EditorConfig/EnableEditorConfigSupport/@EntryValue">False</s:Boolean>
```
Comment (lucaspimentel) [inline Datadog.Trace.Managed.sln.DotSettings]: I don't think we want to disable EditorConfig support. It's the best way to keep settings like these shared across editors and platforms. _This_ file is secondary because it only applies to Resharper and Rider.

File: Datadog.Trace.Managed.sln.DotSettings [hunk 2]
```diff
@@ -54,6 +62,7 @@
 	<s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/AutoDetectedNamingRules/=StaticReadonly/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="" Suffix="" Style="AaBb" /&gt;</s:String>
 	<s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/AutoDetectedNamingRules/=TypeParameters/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="T" Suffix="" Style="AaBb" /&gt;</s:String>
 	<s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/AutoDetectedNamingRules/=TypesAndNamespaces/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="" Suffix="" Style="AaBb" /&gt;</s:String>
+	<s:String x:Key="/Default/CodeStyle/Naming/CSharpNaming/PredefinedNamingRules/=PrivateStaticReadonly/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="_" Suffix="" Style="aaBb" /&gt;</s:String>
```
Comment (lucaspimentel) [inline Datadog.Trace.Managed.sln.DotSettings]: The naming convention for both `static readonly` and `const` fields is Pascal case, even if they are private.

File: Datadog.Trace.Managed.sln.DotSettings [hunk 3]
```diff
@@ -95,6 +104,20 @@
 	<s:String x:Key="/Default/CodeStyle/Naming/XamlNaming/UserRules/=NAMESPACE_005FALIAS/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="" Suffix="" Style="aaBb" /&gt;</s:String>
 	<s:String x:Key="/Default/CodeStyle/Naming/XamlNaming/UserRules/=XAML_005FFIELD/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="" Suffix="" Style="AaBb" /&gt;</s:String>
 	<s:String x:Key="/Default/CodeStyle/Naming/XamlNaming/UserRules/=XAML_005FRESOURCE/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="" Suffix="" Style="AaBb" /&gt;</s:String>
+	<s:String x:Key="/Default/Environment/PerformanceGuide/SwitchBehaviour/=AutoRecoverer/@EntryIndexedValue">LIVE_MONITOR</s:String>
```
Comment (lucaspimentel) [inline Datadog.Trace.Managed.sln.DotSettings]: Let's keep this file to code style (and maybe code inspection) only, please. Developers can customize other IDE and R# settings as they please.

File: integrations.json [hunk 4]
```diff
@@ -153,6 +153,25 @@
       }
     ]
   },
+  {
+    "name": "AspNetWebForms",
```
Comment (lucaspimentel) [inline integrations.json]: Just for naming purposes, we should test this integration with ASP.NET MVC as well. I think it works at a low enough level (ASP.NET) that it applies to any web framework built on top of it. If it does, we could call it simply `AspNet`.

File: Datadog.Trace.Managed.sln.DotSettings (hunk 1 above)
Comment (boydc7) [reply Datadog.Trace.Managed.sln.DotSettings]: Yea, I think a lot of these were supposed to be in my local user file and not this one, I'll fix em up

File: integrations.json (hunk 4 above)
Comment (boydc7) [reply integrations.json]: Yea, it can be certainly, but we'll have to figure out how to differentiate the OperationName ... I'm sure there's a way, but I couldn't come up with one quickly/easily.  i.e. if you look at the actually integration (AspNetWebFormsIntegration), you'll see that it just uses a AspNetWebFormsHttpModule as the IHttpModule to inject, which is a simple implementation of the base AspNetHttpModule.  The only reason that exists right now is to figure out what the OperationName should be (in this case aspnet-web-forms.request).  There'd have to be a way for us to figure out that the integration was actually a web-forms request vs. something else... at least, I think that's the case

File: Datadog.Trace.Managed.sln.DotSettings (hunk 2 above)
Comment (boydc7) [reply Datadog.Trace.Managed.sln.DotSettings]: Alright, I changed it back...I imagine then we should bite the bullet and run a formatter over existing code to apply this stuff, since I just sample a dozen locations where this was not the case...not sure I can find a place where a static readonly actually follows this not counting the ones I just updated...maybe I'm just seeing static that are not readonly, but I'm not sure I see why there'd be a difference...that said, I really don't care, just need to know what it should be so we can be consistent about it

File: integrations.json (hunk 4 above)
Comment (lucaspimentel) [reply integrations.json]: TL;DR: drop references to Web Forms. Use `AspNetIntegration` and `aspnet.request`.

The way I see it, this isn't really a "ASP.NET Web Forms" integration but an ASP.NET integration (as in, the lower-level web framework that integrates with IIS and other web frameworks are built upon) which instrument Web Forms request, among others. We should use a generic name that doesn't mention Web Forms, like `AspNetIntegration` for the integration class and `aspnet.request` for the operation name, and let specific web frameworks add their particular spans as children.

Integration | Operation name | Instrumented type
---|---|---
ASP.NET (this one) | `aspnet.request` | `System.Web.HttpApplication` (to add a custom `System.Web.IHttpModule`)
ASP.NET MVC | `aspnet-mvc.request` | `System.Web.Mvc.Async.IAsyncActionInvoker`
ASP.NET Web API | `aspnet-web-api.request` (should be `aspnet-webapi.request`) | `System.Web.Http.Controllers.IHttpController`
ASP.NET Web Forms (future) | `aspnet-webforms.request` | probably `System.Web.UI.Page`
WCF (future) | `wcf.request` | TBD
Nancy (future) | `nancy.request` | TBD

Something similar would happen in .NET Core where "ASP.NET Core MVC" and other libraries are implemented as middleware on top of "ASP.NET Core".

File: Datadog.Trace.Managed.sln.DotSettings (hunk 2 above)
Comment (lucaspimentel) [reply Datadog.Trace.Managed.sln.DotSettings]: Yeah, you're not wrong. @bmermet used lower camel case for these, so you'll see a lot of them in the repo. I use upper camel case in new ones going forward (see [Tracer.cs](https://github.com/DataDog/dd-trace-dotnet/blob/master/src/Datadog.Trace/Tracer.cs)), and will usually change the ones I see in a file if I am making other changes there. Otherwise I leave them alone to avoid creating lots of unrelated diffs in PRs. I don't think it's worth going back and changing all of them if we will probably catch them all over time.

StyleCop isn't throwing errors on these, is it?

File: integrations.json (hunk 4 above)
Comment (boydc7) [reply integrations.json]: done, makes sense naturally once it's laid out.  So I guess this really isn't so much WebForms integration as it is basic ASP.NET, which also is correct.

While I was in there, I noticed the currently WebApi integration operation name is:

aspnet-web-api.request

Should that be changed to just:

aspnet-webapi.request

?

Comment (lucaspimentel) [reply integrations.json]: > Should that be changed to just: aspnet-webapi.request

Yes, please!

File: src/Datadog.Trace.ClrProfiler.Managed/Datadog.Trace.ClrProfiler.Managed.csproj [hunk 5]
```diff
@@ -1,7 +1,7 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks Condition="'$(OS)' == 'Windows_NT'">netstandard2.0;net45;net461</TargetFrameworks>
+    <TargetFrameworks>net45;net461;netstandard2.0</TargetFrameworks>
```
[the remark this answers is not in the document]
Comment (boydc7) [reply src/Datadog.Trace.ClrProfiler.Managed/Datadog.Trace.ClrProfiler.Managed.csproj]: If you mean the piece of moving netstandard2.0 to the end vs. the beginning, yes (had something to do with getting it to compile/run for webforms samples/site - there's something with the default being the first and what not).  The removing the equals condition on OS platform I can't say would have been on purpose, but either way I've just added it back and re-shuffled the order again (since we likely won't be working with WebForms directly much in the future hopefully).

File: src/Datadog.Trace.ClrProfiler.Managed/DynamicMethodBuilder.cs [hunk 6]
```diff
@@ -183,13 +183,17 @@ public static class DynamicMethodBuilder<TDelegate>
                 dynamicMethod.CallVirtual(methodInfo);
             }
 
-            if (methodInfo.ReturnType.IsValueType && returnType == typeof(object))
+            // Non-void return type?
+            if (returnType != null)
```
[the remark this answers is not in the document]
Comment (boydc7) [reply src/Datadog.Trace.ClrProfiler.Managed/DynamicMethodBuilder.cs]: Yep, agreed

File: src/Datadog.Trace.ClrProfiler.Managed/ExtensionMethods/ObjectExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,19 @@
+using System;
+
+namespace Datadog.Trace.ClrProfiler.ExtensionMethods
+{
+    internal static class ObjectExtensions
+    {
+        internal static void TryDispose(this IDisposable disposable)
```
[the remark this answers is not in the document]
Comment (boydc7) [reply src/Datadog.Trace.ClrProfiler.Managed/ExtensionMethods/ObjectExtensions.cs]: Agreed - sure would be nice if the Scope was stateless and disposal (and other concerns) were handled by a delegate that allowed me to easily handle that in different ways at design/runtime without having to modify existing code and trace through a bunch of locations that shouldn't care (for example, on Close(), should an exception on the scope manager impact or be independent from finishing the span...)...but I digress.  In the interest of getting this done for now, just try/catching there...but there should be consideration given to the existing concerns in Close(), as well as future considerations, and likely some logging of sorts (which would be easy to address if concerns were separated)....

File: src/Datadog.Trace.ClrProfiler.Managed/Datadog.Trace.ClrProfiler.Managed.csproj (hunk 5 above)
[the remark this answers is not in the document]
Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed/Datadog.Trace.ClrProfiler.Managed.csproj]: Makes sense. I think I've hit that issue a few times with the first one being the default. Didn't realize this was causing it until now! Feel free to change the order of target frameworks if that helps.

The OS platform check we need because we can't build `net45` or `net461` on Linux ([yet 🔗](https://github.com/dotnet/designs/blob/1c2baaf1773e52e5c13f94b8659de377c1f16671/accepted/targeting-packs/targeting-packs.md)).

File: src/Datadog.Trace.ClrProfiler.Managed/Services/CompositeSpanDecorator.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
+using System.Collections.Generic;
+using Datadog.Trace.ClrProfiler.Interfaces;
+using Datadog.Trace.ExtensionMethods;
+using Datadog.Trace.Interfaces;
+
+namespace Datadog.Trace.ClrProfiler.Services
+{
+    internal class CompositeSpanDecorator : ISpanDecorator
```
[the remark this answers is not in the document]
Comment (boydc7) [reply src/Datadog.Trace.ClrProfiler.Managed/Services/CompositeSpanDecorator.cs]: Framework doesn't seem to be able to make up it's mind in this case, though from what I can see the use of "Aggregate" is used in many a scenario (aggregate patterns that are not composites, aggregate objects that are accumulators, and aggregates that are in fact composites), whereas "Composite" is used nearly always for a composite pattern implementation (with a few exceptions).  A simple search here for "Aggregate" or "Composite" can give some insight:

https://docs.microsoft.com/en-us/dotnet/api/index?view=netframework-4.7.2&term=composite
https://docs.microsoft.com/en-us/dotnet/api/index?view=netframework-4.7.2&term=aggregate

An aggregate isn't a composite (and vice versa).   Aggregate exception is a bad example, as it is not a composite (composite patterns are for abstracting an exact operation over an unknown number of concrete implementations (or a single concrete implementation that isn't a composite) interchangeably without any knowledge of/concern about the underlying implementation.  An aggregate object (i.e. an aggregate exception) is an Exception and can be used as one, but only exposes additional state (i.e. a collection of inner-exceptions) when you explicitly code to that (i.e. cast/code to an AggregateException to get the InnerExceptions collection).  Akin to an accumulator.)

There certainly are some framework objects that are named Aggregate but are composites (AggregateCacheDependency as an example), but there are many a many that are either aggregate objects or accumulators (i.e. AggregateException or the Aggregate extension).

Long story short for me - I'd be as explicit about what we're doing as possible, and we're not aggregating/accumulating things...we're composing them into functionality.

[the remark this answers is not in the document]
Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed/Services/CompositeSpanDecorator.cs]: Sounds good. Thanks for doing the research!

File: src/Datadog.Trace.ClrProfiler.Managed/GlobalSuppressions.cs [hunk 9]
```diff
@@ -4,3 +4,4 @@
...
 [assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("StyleCop.CSharp.MaintainabilityRules", "SA1119:Statement must not use unnecessary parenthesis", Justification = "ActuallyNeeded", Scope = "member", Target = "~M:Datadog.Trace.ClrProfiler.Integrations.AspNetHttpModule.TryGetContext(System.Object,System.Web.HttpContext@)~System.Boolean")]
 [assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("StyleCop.CSharp.MaintainabilityRules", "SA1119:Statement must not use unnecessary parenthesis", Justification = "ActuallyNeeded", Scope = "member", Target = "~M:Datadog.Trace.ClrProfiler.Integrations.AspNetIntegration.Init(System.Object)")]
+[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("StyleCop.CSharp.ReadabilityRules", "SA1118:Parameter must not span multiple lines", Justification = "ItShould", Scope = "member", Target = "~M:Datadog.Trace.ClrProfiler.Services.DependencyFactory.#cctor")]
```
Comment (lucaspimentel) [inline src/Datadog.Trace.ClrProfiler.Managed/GlobalSuppressions.cs]: `GlobalSuppressions.cs` is meant for... well, global suppressions 🙂. The ones you have added have a specific target (`~M:Datadog.Trace.ClrProfiler.Services.DependencyFactory.#cctor`), so you added `SA1119` several times. If you want them to be really global, don't set the `Target` property. If your intention was to disable the check in a specific method, you can set the attribute on that method directly instead of doing it at the assembly level. Note also that we have shared cascading `GlobalSuppressions.cs` files, so it's easy to apply suppressions across multiple projects.

File: test/Datadog.Trace.ClrProfiler.Managed.Tests/Datadog.Trace.ClrProfiler.Managed.Tests.csproj [hunk 10]
```diff
@@ -10,4 +10,10 @@
     <ProjectReference Include="..\Datadog.Trace.TestHelpers\Datadog.Trace.TestHelpers.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Reference Include="System.Web">
+      <HintPath>..\..\..\..\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.6.1\System.Web.dll</HintPath>
```
Comment (lucaspimentel) [inline test/Datadog.Trace.ClrProfiler.Managed.Tests/Datadog.Trace.ClrProfiler.Managed.Tests.csproj]: Having the `<HintPath>` here is weird, especially since it includes a specific framework version. If you had to add this so the `netcoreapp2.1` wouldn't fail, it's because `System.Web` should only be included when building for .NET Framework and not for .NET Core or .NET Standard:

```xml
<ItemGroup Condition=" '$(TargetFramework)' != 'netcoreapp2.1' ">
    <Reference Include="System.Web" />
</ItemGroup>
```

File: src/Datadog.Trace/ExtensionMethods/DictionaryExtensions.cs [hunk 11]
```diff
@@ -13,5 +15,33 @@ internal static class DictionaryExtensions
 
             return default(TValue);
         }
+
+        public static bool TryGetValueOrDefaultAs<TAs>(this IDictionary dictionary, object key, out TAs valueAs)
```
[the remark this answers is not in the document]
Comment (boydc7) [reply src/Datadog.Trace/ExtensionMethods/DictionaryExtensions.cs]: The As is because it's being casted to something else (this is a non-generic IDictionary of objects (i.e. a HashTable)).  If the key doesn't exist, you'll get different exceptions/results depending on the concrete implementation (HashTable I believe is just null returned, whereas more modern implementers throw KeyNotFound and/or NotSupported exceptions).

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetHttpModule.cs [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+                Log.ErrorException("Datadog ASP.NET HttpModule instrumentation error", ex);
+            }
+        }
+
+        private void OnEndRequest(object sender, EventArgs eventArgs)
+        {
+            try
+            {
+                if (!TryGetContext(sender, out var httpContext) ||
+                    !httpContext.Items.TryGetValueOrDefaultAs<Scope>(_httpContextKey, out var scope))
```
[the remark this answers is not in the document]
Comment (boydc7) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetHttpModule.cs]: Only if the concrete implementation is a HashTable or similar...those aren't the only IDictionary implementations

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetIntegration.cs [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+        /// </summary>
+        /// <param name="thisObj">The HttpApplication instance ref.</param>
+        [InterceptMethod(
+            TargetAssembly = "System.Web",
+            TargetType = "System.Web.HttpApplication")]
+        public static void Init(object thisObj)
+        {
+            try
+            {
+                var initMethodAction = DynamicMethodBuilder<Action<object>>.GetOrCreateMethodCallDelegate(thisObj.GetType(), "Init");
```
[the remark this answers is not in the document]
Comment (boydc7) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetIntegration.cs]: I used the IL as I can only actually call the Init() directly if the cast is successful.  Maybe we don't care if the cast isn't successful I suppose (i.e. just let it throw an exception), but that didn't seem right to me.  That said, I can easily obviously move the IL emt into that portion of the if block also

[the remark this answers is not in the document]
Comment (boydc7) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetIntegration.cs]: I'm not so sure about the ordering, as the only example I can find orders it the same way, even though it does seem odd (i.e.init the application first, module 2nd).

File: src/Datadog.Trace/ExtensionMethods/DictionaryExtensions.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (lucaspimentel) [reply src/Datadog.Trace/ExtensionMethods/DictionaryExtensions.cs]: Oh, I forgot to remove this `As` comment after we spoke about it over the phone. Feel free to ignore it.

> If the key doesn't exist, you'll get different exceptions/results depending on the concrete implementation

Not if you're using the older non-generic `IDictionary`. [`IDictionary` always returns null when a key is not found](https://docs.microsoft.com/en-us/dotnet/api/system.collections.idictionary.item). In .NET 1.x where there was no way to distinguish between a missing key and an existing key with a null value (which is valid). In .NET 2.0, they decided to fix this by having `IDictionary<TKey, TValue>` throw an exception unless you use `TryGetValue()`, but changing the old `IDictionary` would be a breaking change so we're stuck with that behavior forever.

For example, `Dictionary<TKey, TValue>` (which implements both interfaces) throws an exception when used directly or via the generic interface, but it will return null when used via the non-generic interface. This is intentional to avoid breaking the `IDictionary` contract.

```csharp
var foo = new Dictionary<string,string>();
var value1 = foo ["key"]; // throws exception

IDictionary bar = foo;
var value2 = bar["key"]; // same reference, but returns null
```

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetHttpModule.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetHttpModule.cs]: Any implementation of [`IDictionary.Item[object]`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.idictionary.item) they throws instead of returning null on missing key is broken.

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetIntegration.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/AspNetIntegration.cs]: Whenever possible, we should let the C# compiler generate the IL for us instead of writing it by hand at run time and calling it through a delegate (but avoid using `dynamic` because that's actually worse). 

C# compiler > emit IL > `dynamic` > plain old reflection

Also, if the cast to `HttpApplication` ever fails, we have bigger problems to worry about 😉
